src/crypto/sha256.cpp

```diff
@@ -716,6 +716,60 @@ void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned
     }
 }
 
+std::vector<unsigned char> CSHA256::Save() const {
+    size_t buf_size = bytes % 64;
+    std::vector<unsigned char> result(40 + buf_size);
+
+    WriteBE32(&result[ 0], s[0]);
+    WriteBE32(&result[ 4], s[1]);
+    WriteBE32(&result[ 8], s[2]);
+    WriteBE32(&result[12], s[3]);
+    WriteBE32(&result[16], s[4]);
+    WriteBE32(&result[20], s[5]);
+    WriteBE32(&result[24], s[6]);
+    WriteBE32(&result[28], s[7]);
+
+    WriteLE64(&result[32], bytes << 3);
+
+    memcpy(&result[40], buf, buf_size);
+
+    return result;
+}
+
+bool CSHA256::Load(const std::vector<unsigned char>& vch) {
+    if (vch.size() < 40) return false;
+
+    uint64_t bits = ReadLE64(&vch[32]);
+    size_t buf_size = (bits >> 3) % 64;
+
+    if ((bits & 0x07) != 0 || vch.size() != 40 + buf_size) return false;
```

Comment:
Maybe add a comment to state that bits & 0x07 check is to prevent malleability ?

Reply:
It's a bit more than that: the `SHA256` class is simply incapable of processing sub-byte data as it stands.  But I can write a comment to that effect.

Reply:
I mean you could just just do `bits &= ~0x07` and not do the check, but this would introduce non-determinism, and the comment could tell the reader why it is not a good idea

Reply:
I had hoped it goes without saying that one shouldn't mangle an invalidly serialized data blob into a valid one.

Reply:
validity of serialized data depends on the definition of the validity, but it is not explicitly specified (the serialization code in `Save()` is only one possible implementation of implicit specification). But, on the other hand, you can view the `(bits & 0x07) != 0` as the statement of the de-facto specification "low-order bits should be zero", and in this light, no extra comment seems necessary.